Storage/container/one2one.h

```diff
@@ -0,0 +1,253 @@
+  struct MapAccessor final {
+    using T_KEY = typename T_MAP::key_type;
+    const T_MAP& map_;
+
+    struct Iterator final {
+      using T_ITERATOR = typename T_MAP::const_iterator;
+      T_ITERATOR iterator_;
+      explicit Iterator(T_ITERATOR iterator) : iterator_(iterator) {}
+      void operator++() { ++iterator_; }
+      bool operator==(const Iterator& rhs) const { return iterator_ == rhs.iterator_; }
+      bool operator!=(const Iterator& rhs) const { return !operator==(rhs); }
+      sfinae::CF<T_KEY> key() const { return iterator_->first; }
+      const T& operator*() const { return *iterator_->second; }
+      const T* operator->() const { return iterator_->second; }
+    };
+
+    explicit MapAccessor(const T_MAP& map) : map_(map) {}
+
+    bool Empty() const { return map_.empty(); }
+
+    size_t Size() const { return map_.size(); }
+
+    bool Has(const T_KEY& x) const { return map_.find(x) != map_.end(); }
+
+    Iterator begin() const { return Iterator(map_.cbegin()); }
+    Iterator end() const { return Iterator(map_.cend()); }
+  };
+
+  const MapAccessor<T_FORWARD_MAP> Rows() const { return MapAccessor<T_FORWARD_MAP>(forward_); }
+
+  const MapAccessor<T_TRANSPOSED_MAP> Cols() const { return MapAccessor<T_TRANSPOSED_MAP>(transposed_); }
+
+  struct Iterator final {
+    using T_ITERATOR = typename T_WHOLE_MAP::const_iterator;
+    T_ITERATOR iterator_;
+    explicit Iterator(T_ITERATOR iterator) : iterator_(iterator) {}
+    void operator++() { ++iterator_; }
+    bool operator==(const Iterator& rhs) const { return iterator_ == rhs.iterator_; }
+    bool operator!=(const Iterator& rhs) const { return !operator==(rhs); }
+    const T_RCPAIR key() const { return iterator_->first; }
+    const T& operator*() const { return *iterator_->second; }
+    const T* operator->() const { return iterator_->second; }
+  };
+  Iterator Begin() const { return Iterator(map_.begin()); }
+  Iterator End() const { return Iterator(map_.end()); }
+
+ private:
+  void DoErase(const T_RCPAIR& row_col) {
+    forward_.erase(row_col.first);
+    transposed_.erase(row_col.second);
+    map_.erase(row_col);
+  }
+
+  void DoAdd(const T_RCPAIR& row_col, const T& object) {
+    auto& placeholder = map_[row_col];
+    placeholder = std::make_unique<T>(object);
+    forward_[row_col.first] = placeholder.get();
+    transposed_[row_col.second] = placeholder.get();
+  }
+
+  T_WHOLE_MAP map_;
+  T_FORWARD_MAP forward_;
+  T_TRANSPOSED_MAP transposed_;
+  MutationJournal& journal_;
+};
+
+template <typename T, typename T_UPDATE_EVENT, typename T_DELETE_EVENT>
+using UnorderedOne2One = GenericOne2One<T, T_UPDATE_EVENT, T_DELETE_EVENT, Unordered, Unordered>;
+
+template <typename T, typename T_UPDATE_EVENT, typename T_DELETE_EVENT>
+using OrderedOne2One = GenericOne2One<T, T_UPDATE_EVENT, T_DELETE_EVENT, Ordered, Ordered>;
+
+}  // namespace container
+
+template <typename T, typename E1, typename E2>  // Entry, update event, delete event.
```

Comment:
OK. I was thinking of some sort of code completion that suggests E1 while it could suggest what's really should be passed in there. The detailed compilation errors would also be more readable with semantic names. My $0.02, probably you've got other use cases in mind.